### PR TITLE
Truncate file

### PR DESCRIPTION
**Pull request by erssebaggala**
Truncate file to handle cases where the write index ends up behind the read index. This occurs in cases where the content being written is less than what is being read. As a consequence, you end up with extra unwanted data from the original file.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -95,14 +96,18 @@ module.exports = (async (file, {readStart, writeStart} = {}) => {
...
-				})
+                .then(() => close(fd))
                 .then(callback);
         }
     });
 
+	
+	//Truncate the file up to where the write ends, otherwise
+	//some of the original content will remain in place
+	writeStream.on('finish', () => {
```
Comment (MichalCz) [inline index.js:107]: Hi,

I'm not sure if it makes sense to move the truncate here. Writing may not yet be fully done. The prior option (truncate in flush) would be better, but it should be something like:

```javascript
_writePromise
.then(() => fs.ftruncate(fd, writeIndex))
.then(() => close(fd))
.then(() => callback())
.catch(e => callback(e))
```

This way we make sure the truncate runs AFTER the file is written and we ommit synchronous FS.

**Review by MichalCz — CHANGES_REQUESTED** [review]

File: index.js (hunk 1 above)
Comment (MichalCz) [reply index.js:107]: Hey, I'll fix this during the weekend.

I'll need to add some test scenarios also.

Comment (erssebaggala) [reply index.js:107]: Hi, I agree. However, I had  noticed that **flush** is never called. I checked [ _stream_writeable.js ](https://github.com/nodejs/node/blob/64219741218aa87e259cf8257596073b8e747f0a/lib/_stream_writable.js) and I don't see any reference to **flush**.  I am working with node@v10.15.3.

Comment (erssebaggala) [reply index.js:107]: The only reference to **flush** in the docs is in the transform stream which is called when there is no more data to consume. I could be wrong but it looks like we should not be implementing flush in  the writeable as doing so has no effect!

**Comment by MichalCz** [comment]
You may be right. Perhaps out of routine I thought that.

I'll check it today.

sob., 8 cze 2019, 01:47 użytkownik Emmanuel Robert Ssebaggala <
notifications@github.com> napisał:

**Comment by MichalCz** [comment]
Hi @erssebaggala - I've merged this to develop manually - actually made the fixes you mentioned in `writable._final` where they belong.

Now it should work as expected. Please let me know if this helped.

**Comment by erssebaggala** [comment]
@MichalCz, I will let you know after testing. Thanks

**Pull request closed without merging**